baselines/jft/active_learning.py

```diff
@@ -0,0 +1,450 @@
+
+  # TODO: val_ds for early stopping
+
+  test_ds = input_utils.get_data(
+    dataset=config.dataset,
+    split="test",
+    rng=None,
+    host_batch_size=local_batch_size_eval,
+    preprocess_fn=preprocess_spec.parse(
+      spec=config.pp_eval, available_ops=preprocess_utils.all_ops()
+    ),
+    shuffle=False,
+    prefetch_size=config.get("prefetch_to_host", 2),
+    num_epochs=1,  # Only repeat once.
+  )
+
+  model = ub.models.vision_transformer(
+      num_classes=config.num_classes, **config.get("model", {})
+  )
+
+  rng, rng_init = jax.random.split(rng)
+
+  init = make_init(model, test_ds, local_batch_size, config)
+  params_cpu = init(rng_init)
+
+  # Load the optimizer from flax.
+  opt_name = config.get("optim_name")
+  opt_def = getattr(flax.optim, opt_name)(**config.get("optim", {}))
+
+  # We jit this, such that the arrays that are created are created on the same
+  # device as the input is, in this case the CPU. Else they'd be on device[0].
+  opt_cpu = jax.jit(opt_def.create)(params_cpu)
+
+  reinit_params = config.get("model_reinit_params", ("head/kernel", "head/bias"))
+  loaded = checkpoint_utils.load_from_pretrained_checkpoint(
+    params_cpu,
+    config.model_init,
+    config.model.representation_size,
+    config.model.classifier,
+    reinit_params,
+  )
+
+  opt_cpu = opt_cpu.replace(target=loaded)
+
+  # TODO: This shouldn't be needed but there are issues with opt_cpu being donated
+  opt_cpu = jax.device_get(opt_cpu)
+
+  update_fn = make_update_fn(model, config)
+  evaluation_fn = make_evaluation_fn(model, config)
+
+  acquisition_batch_size = 10
+
+  # NOTE: if we could `enumerate` before `filter` in `create_dataset` of CLU then this dataset creation could be simplified
+  # https://github.com/google/CommonLoopUtils/blob/main/clu/deterministic_data.py#L340
+  # CLU is explicitly not accepting outside contributions at the moment.
+  train_subset_data_builder = tfds.builder(
+      "cifar10_subset", subset_ids={"train": [], "test": None}
+  )
+  train_subset_data_builder.download_and_prepare()
+
+  pool_subset_data_builder = tfds.builder(
+      "cifar10_subset", subset_ids={"train": None, "test": None}
+  )
+
+  rng, pool_ds_rng = jax.random.split(rng)
+
+  # NOTE: below line is necessary on multi host setup
+  # pool_ds_rng = jax.random.fold_in(pool_ds_rng, jax.process_index())
+
+  pool_train_ds = input_utils.get_data(
+    dataset=pool_subset_data_builder,
+    split=config.train_split,
+    rng=pool_ds_rng,
+    host_batch_size=local_batch_size,
+    preprocess_fn=preprocess_spec.parse(
+        spec=id_pp_eval, available_ops=preprocess_utils.all_ops()
+    ),
+    shuffle=False,
+    drop_remainder=False,
+    prefetch_size=config.get("prefetch_to_host", 2),
+    num_epochs=1,  # Don't repeat
+  )
+
+  test_accuracies = []
+  training_size = []
+
+  rng, rng_loop = jax.random.split(rng, 2)
+  rngs_loop = flax_utils.replicate(rng_loop)
+
+  while True:
```

Comment:
Updating to `while current_train_ds_length < 150` would be easier to quickly understand.

Reply:
What's our target Python version at Google? Can we use the `:=` syntax yet?

IIRC the reason for this was just so that `current_train_ds_length` is only assigned once

Reply:
We should target 3.7 compatibility, so can't use `:=` yet.